Maxgamill sheffield/692 standardised images

MaxGamill-Sheffield opened the pull request:
closes #692 

Adds:
- a matplotlib stylesheet incase of more modifications
- larger font sizes (18pt - 16pt instead of 10pt) for images
- function (and test) & config option to change the number of ticks on the axes
- config documentation update

I've left out using the style sheet for plotting graphs as these may need another session with the TopoStats users to see what plots they want and how they want these to look etc.

ns-rse commented on topostats/plottingfuncs.py:
```diff
@@ -93,6 +94,7 @@ def __init__(
         zrange: list = None,
         colorbar: bool = True,
         axes: bool = True,
+        num_ticks=None,
```
```suggestion
        num_ticks:int = None,
```

MaxGamill-Sheffield replied:
This but using union of int and None

ns-rse reviewed (changes requested):
Thanks for picking this up @MaxGamill-Sheffield looks good.

I think for flexibility we should make every option that we are setting in the Matplotlib style a configurable option and vice-versa any option should then be explicitly defined in the style. My thinking being that at some point someone is going to want to change them (e.g. users outside of the group). What do you think?

MaxGamill-Sheffield commented on topostats/images.mplstyle:
```diff
@@ -0,0 +1,2 @@
+font.size: 16
+axes.titlesize: 18
```
So the way the ticks and values are set is really unintuitive and uses "Locators". As such this is not something that is standardised within mpl and thus configurable the stylesheets in the stylesheet (unless we can pull parameters from the style sheet for use in the file). Hence why this option is configurable in the config (although I don't like it there as I think it clutters the config so maybe moving plotting config parameters to their own file in another PR might be better)

From the session we had it was mainly this and the colourmap but those edits are going to be a seperate PR when we can pick what the group likes best. Although I've used a stylesheet so if we want to add more later we can easily. Also see Alice's point of making a "brand".

ns-rse replied:
Ok, I'm kind of thinking beyond the group though as part of the plan is to attract a wider user base and other groups or individuals are likely to have their own preferences.

I'm yet to be convinced that splitting configuration across multiple files actually helps reduce complexity in anyway, at best it obfuscates it, at worse it makes it more confusing/complex having to look in different files.

MaxGamill-Sheffield commented on topostats/plottingfuncs.py:
```diff
@@ -18,6 +18,7 @@
 # pylint: disable=dangerous-default-value
 
 LOGGER = logging.getLogger(LOGGER_NAME)
+plt.style.use(Path(__file__).resolve().parent / "images.mplstyle")
```
Dm sylvia and I found a way using:
`plt.style.use(pkg_resources.files(topostats) / "images.mplstyle")`

MaxGamill-Sheffield commented on topostats/plottingfuncs.py:
```diff
@@ -136,6 +138,8 @@ def __init__(
             Optionally add a colorbar to plots, default is False.
         axes: bool
             Optionally add/remove axes from the image.
+        num_ticks: int
```
Fixed, although the object passed is a tuple as it seems the conversion using a tuple in the config file doesn't work (maybe there needs to be some safe parsing option used)

MaxGamill-Sheffield commented on topostats/default_config.yaml:
```diff
@@ -63,6 +63,7 @@ plotting:
   zrange: [null, null]  # low and high height range for core images (can take [null, null]). low <= high
   colorbar: true  # Options : true, false
   axes: true # Options : true, false (due to off being a bool when parsed)
+  num_ticks: null # Number of ticks to have along the xy axes. Options : null (auto) or integer
```
Hey Neil, please could you explain this further? This is already overridable and in the config so I don't understand what you mean.

ns-rse replied:
Sorry for not being clearer.

In my view everything in the style sheet should also be configurable, thus the `font.size` and `axis.titlesize` should be in the configuration so that users outside of the group can change them.

Ultimately anyone can either extract the data they want to plot and do so manually or hack at the code.

Good software gives people freedom and doesn't impose constraints on them.

MaxGamill-Sheffield commented:
Added:
- List input format for [x, y] num_ticks
- Updated validation and typehints accordingly
- set_n_ticks() now handles the None's
- Cleaned "valued" to "values" in validation
- Stricter validations on zrange - list -> [int, float, None]

SylviaWhittle commented:
Looking good! Posting an image here to show what the default output looks like now 😄 
![image](https://github.com/AFM-SPM/TopoStats/assets/86117496/4fbf6994-35b6-4a20-96d5-7033f67bc9ea)

ns-rse reviewed (approved):
Thanks for addressing some of the suggestions @MaxGamill-Sheffield, given we handle non-square sans I thought it prudent to support different numbers of ticks on each axis.

Having a set of defaults that are used but also making them configurable are not incompatible. One of the historical problems with the code base that we have overcome is that there were hard-coded variables that required people to hack at the code resulting in a poor user experience.

I've put some thoughts as to a way forward down in #704, views and ideas welcome.
